### PR TITLE
MCLOUD-5991: Restore additional data in env.php if this file was removed

### PR DESCRIPTION
**Pull request by NadiyaS**
### Description
If env.php does not exist and DB is full of data, ece-tools runs setup:upgrade, but the deployment is still failed. 

- check existence of crypt key. If this data does not exist (whether in env.php or in .magento.env.yaml) - throw exception that there is not enough data for setup:upgrade or DB is not empty for setup:install and provide appropriate steps to resolve this. Otherwise, continue to run Magento upgrade process as it exists now.
- if env.php did not exist or had only one cache type in it - restore all default types, otherwise keep them as they are (we cannot restore cache type data by default as customer could remove some types for purpose)
- add warning that env.php did not exist at the beginning of deployment.

### Fixed Issues (if relevant)

1. https://jira.corp.magento.com/browse/MCLOUD-5991

### Manual testing scenarios

1. on installed Magento remove env.php and run redeployment 
**Expected result:** deployment is failed with message that `crypt key` does not exist
2. remove env.php again (as it was partially restored form prev. deployment) add CRYPT_KEY variable and run redeployment
**Expected result:**
- deployment finished without errors
- env.php has whole list of `cache_types`
- cloud.log contain message that cache_types was restored and that env.php did not exist at the beginning of the deployment
3. in env.php remove a few types in `cache_type` option and set `0` to other cache_types and left few with value `1`. Run redeployment.
**Expected result:** deployment finished. env.php has the same state of `cache_types` option (few removed types, and types with values as 0 and 1)

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Review by oshmyheliuk — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

File: src/Filesystem/Flag/Manager.php
Comment (bbatsche) [inline src/Filesystem/Flag/Manager.php:43]: My understanding was that the flags were useful to communicate between build & deploy stages, or between multiple deployments. Why are we using them here were everything is happening within a single deployment?

File: src/Step/Deploy/InstallUpdate/ConfigUpdate/CacheType.php
Comment (bbatsche) [inline src/Step/Deploy/InstallUpdate/ConfigUpdate/CacheType.php]: If I understand your comment correctly, this exception will change slightly once specific error codes are created? Otherwise it seems like we should be able to just catch `GenericException`s below.

**Review by bbatsche — CHANGES_REQUESTED** [review]
I have a couple of questions but ultimately the code looks good. I'll approve once they're answered.

File: src/Step/Deploy/InstallUpdate/ConfigUpdate/CacheType.php
Comment (NadiyaS) [reply src/Step/Deploy/InstallUpdate/ConfigUpdate/CacheType.php]: yes, as the codes for theres exceptions will be different I caught both types, just not to forget about it after merging PR https://github.com/magento/ece-tools/pull/726/, e.g. https://github.com/magento/ece-tools/pull/726/files#diff-11ebfae7dd9722de4d90062c73b5170fR89

File: src/Filesystem/Flag/Manager.php
Comment (NadiyaS) [reply src/Filesystem/Flag/Manager.php:43]: good question. Actually, we can understand whether env.php was removed or not only at the beginning of deploy phase, because already on the second step `cache` https://github.com/magento/ece-tools/blob/develop/scenario/deploy.xml#L8 we write data to env.php file and if it is absent we create this file with some part of data. But steps which should react on file absence go much later and at that moment file always exist

**Review by bbatsche — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Review by oshmyheliuk — DISMISSED** [review]

**Review by bbatsche — DISMISSED** [review]

**Review by oshmyheliuk — APPROVED** [review]

**Review by bbatsche — APPROVED** [review]

**Comment by andriyShevtsov** [comment]
QA approved

**Pull request merged**
